[FXML-1991] TOSA to XTenNN for QDQ Operations

cmcgirr-amd opened the pull request:
Conversion pass to match and rewrite TOSA operations to XTeNN quantize and dequantize operations.
```
  func.func @forward(%arg0: tensor<32x3x224x224xf32> ) -> tensor<32x3x224x224xf32> attributes {} {
    %18 = "tosa.const"() {value = dense<0> : tensor<1x1x1x1xi8>} : () -> tensor<1x1x1x1xi8> 
    %19 = "tosa.const"() {value = dense<3.125000e-02> : tensor<1x1x1x1xf32>} : () -> tensor<1x1x1x1xf32> 
    %20 = "tosa.reciprocal"(%19) : (tensor<1x1x1x1xf32>) -> tensor<32x3x224x224xf32> 
    %21 = "tosa.mul"(%arg0, %20) {shift = 0 : i32} : (tensor<32x3x224x224xf32>, tensor<32x3x224x224xf32>) -> tensor<32x3x224x224xf32> 
    %22 = "tosa.cast"(%21) : (tensor<32x3x224x224xf32>) -> tensor<32x3x224x224xi8> 
    %23 = "tosa.add"(%22, %18) {} : (tensor<32x3x224x224xi8>, tensor<1x1x1x1xi8>) -> tensor<32x3x224x224xi8> 
    %24 = "tosa.sub"(%23, %18) {} : (tensor<32x3x224x224xi8>, tensor<1x1x1x1xi8>) -> tensor<32x3x224x224xi8> 
    %25 = "tosa.cast"(%24) : (tensor<32x3x224x224xi8>) -> tensor<32x3x224x224xf32> 
    %26 = "tosa.mul"(%25, %19) {shift = 0 : i32} : (tensor<32x3x224x224xf32>, tensor<1x1x1x1xf32>) -> tensor<32x3x224x224xf32> 
    return %26 : tensor<32x3x224x224xf32> 
  } 
```
To the following:
```
func.func @forward(%arg0: tensor<32x3x224x224xf32>) -> tensor<32x3x224x224xf32> attributes {} {
    %0 = xten_nn.quantize(%arg0 : tensor<32x3x224x224xf32>) {shift = -5 : si32} -> tensor<32x3x224x224xsi8>
    %1 = xten_nn.dequantize(%0 : tensor<32x3x224x224xsi8>) {shift = -5 : si32} -> tensor<32x3x224x224xf32>
    return %1 : tensor<32x3x224x224xf32>
  }
```
This pass also relies on changes to constant folding in TOSA so that the reciprocal, add, and sub operations are folded away.
https://github.com/Xilinx/llvm-project/pull/32
https://github.com/Xilinx/llvm-project/pull/33

The idea in the future would be to generalize these patterns to work on single casts or MULs and fold them into their respective QDQ operations. But for now we restrict the patterns to match exactly on `Q->DQ` as we have a restriction that both the Q and DQ operations must have the same scaling. 
Decoupling the matching would mean that we can have instances where two Q and DQ in theory are matched separately but with different scaling.

Commit:
feat(TosaToXTenNN): partial conversion of TOSA ops to XTenNN Quantization Ops

Commit:
test(quantization.mlir): test cases to matching and rewriting to QDQ operations

Commit:
docs(TosaToXTenNN.cpp): conversion for tosa to XTenNN QDQ

mgehre-amd commented on lib/Conversion/TosaToXTenNN.cpp:
```diff
@@ -0,0 +1,225 @@
+      cast<TensorType>(castOp->getResult(0).getType());
+  return outputTensorType.getElementType().isF32() &&
+         inputTensorType.getElementType().isInteger(8) &&
+         inputTensorType.getElementType().isSignlessInteger();
```
Why restrict it to signless here? I this just temporary?

cmcgirr-amd replied:
Yes simply due to the restriction in TOSA where signless means signed at the moment. I would also love to have `si<>` and `ui<>` in the future. 
For our use case we will restrict ourselves to f32 and i8.

mgehre-amd commented on lib/Conversion/TosaToXTenNN.cpp:
```diff
@@ -0,0 +1,225 @@
+    rewriter.replaceOpWithNewOp<amd::xten_nn::DequantizeOp>(
+        castOp, outputTensorType, newQuantizeOp->getResult(0),
+        /*shift*/ (int32_t)0);
+    rewriter.eraseOp(quantizeOp);
```
This is only safe when quantizeOp had a single user. Can you add a test case where that ops has two users and then fix the code here by checking num_users?

mgehre-amd commented on lib/Conversion/TosaToXTenNN.cpp:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    // We want to make sure we have QDQ operations surrounded by MULs.
+    Operation *mulOperandOp = mulOp->getOperand(0).getDefiningOp();
```
I think you can combine this with line `auto dequantizeOp = cast<amd::xten_nn::DequantizeOp>(mulOperandOp);` below into `auto dequantizeOp = mulOp->getOperand(0).getDefiningOp<amd::xten_nn::DequantizeOp>();`

mgehre-amd commented on lib/Conversion/TosaToXTenNN.cpp:
```diff
@@ -0,0 +1,225 @@
+    }
+    auto dequantizeOp = cast<amd::xten_nn::DequantizeOp>(mulOperandOp);
+
+    // The quantize operation will have it's reciprocal value constant folded.
```
I was expecting that you would compare whether the integer scaleFactor are negated between the first and last mul, but this will also work

cmcgirr-amd replied:
I was in the reciprocal mindset. Let me try if that is simpler to describe.

mgehre-amd commented on lib/Conversion/TosaToXTenNN.cpp:
```diff
@@ -0,0 +1,225 @@
+      return rewriter.notifyMatchFailure(
+          mulOp.getLoc(), "Adding the shifts of the mul and QDQ overflowed.");
+    }
+    quantizeOp.setShift((int32_t)shiftSum.getSExtValue());
```
I don't think that you are allowed to modify ops without telling the rewriter. Typical, people just create a new quanitze op, copying most parameters.
Otherwise you need to use (start|cancel|finalize)RootUpdate, see https://mlir.llvm.org/docs/PatternRewriter/#pattern-rewriter

cmcgirr-amd replied:
I did not know this :D good to know

mgehre-amd commented on lib/Conversion/TosaToXTenNN.cpp:
```diff
@@ -0,0 +1,225 @@
+      return rewriter.notifyMatchFailure(
+          mulOp.getLoc(), "constant is not a integer log2 value.");
+    }
+
```
You will also need to check whether the muls first operand has the same shape as the muls result. Otherwise your transformation is not legal. The shapes could be different when broadcasting happens.

mgehre-amd commented on lib/Conversion/TosaToXTenNN.cpp:
```diff
@@ -0,0 +1,225 @@
+    RewritePatternSet patterns(context);
+
+    // Ensures constants on the add, mul, sub are on the RHS
+    populateCommutativityUtilsPatterns(patterns);
```
Did this work for you? When @TinaAMD and I looked at it, the pass didn't do anything.

cmcgirr-amd replied:
It does for me. I'll add a test case so we see it happening.

mgehre-amd reviewed (changes requested):
Looks good with a few small comments

Commit:
fix(TosaToXTenNN.cpp): add check for single use

multiple use of the cast would mean we cannot convert to QDQ

Commit:
refactor(TosaToXTenNN.cpp): simplify the scale calculation logic

rather than performing 1/scale simply convert both to log2 form and check if they are equal and negation of each other.

Commit:
refactor(TosaToXTenNN.cpp): create new QDQ rather than in place

also renamed the multiplications to show which operation they are associated with

Commit:
fix(TosaToXTenNN.cpp): check input and output shapes on MULs

ensuring broadcasting did not occur

Commit:
test(quantization.mlir): SortCommuntativeOperands test case for QDQ

mgehre-amd reviewed (approved):
Looks good!

mgehre-amd reviewed (approved):
Looks good!

ljfitz commented on include/xten/Conversion/Passes.td:
```diff
@@ -36,4 +37,16 @@ def XTenToLinalg : Pass<"xten-to-linalg", "ModuleOp"> {
 
+def TOSAToXTenNN : Pass<"tosa-to-xten-nn", "ModuleOp"> {
+  let summary = "Partial Conversion of TOSA to XTenNN.";
+  let description = [{Converts all TOSA operators that match to the quanization
```
```suggestion
  let description = [{Converts all TOSA operators that match to the quantization
```

cmcgirr-amd committed:
Update include/xten/Conversion/Passes.td

Co-authored-by: Liam Fitzpatrick <liam.fitzpatrick@xilinx.com>

TinaAMD commented on lib/Conversion/TosaToXTenNN.cpp:
```diff
@@ -0,0 +1,265 @@
+///\return true represents a quantize cast
+///\return false does not represent a quantize cast
+bool isF32ToI8Cast(tosa::CastOp castOp) {
+  TensorType inputTensorType =
```
Maybe you could have one function checking "converts f32/i8" and call it with swapped `inputTensorType`/`outputTensorType` to implement those two functions (very optional).

Commit:
Merge remote-tracking branch 'origin/feature/fused-ops' into christopher.FXML-1991_tosa_to_xten_nn

Commit:
style(XTenNNToTosaPass.h): align filename to the other headers

Commit:
style(quantization.mlir): add newline to file ending

TinaAMD commented on lib/Conversion/TosaToXTenNN.cpp:
```diff
@@ -0,0 +1,265 @@
+/// Used to check that broadcasting did not occur, for example, on
+/// multiplication operations.
+///
+///\param operation we are inspecting. Assumes operand(0) and result(0) exists.
```
exist~s~

TinaAMD commented on lib/Conversion/TosaToXTenNN.cpp:
```diff
@@ -0,0 +1,265 @@
+
+  LogicalResult matchAndRewrite(tosa::MulOp dequantizeMulOp,
+                                PatternRewriter &rewriter) const override {
+
```
Intuitively, I would first check whether the mul is actually operating on a dequantize and bail if this is not the case (i.e. use `dequantizeOp` from line 159 and bail with the error). Seems to be a cheep and obvious reason for failing (even more than the constant).

TinaAMD commented on lib/Conversion/TosaToXTenNN.cpp:
```diff
@@ -0,0 +1,265 @@
+        dequantizeOp->getOperand(0).getDefiningOp());
+
+    // Make sure these multiplications really only belong to the QDQ operations
+    // and used by no one else
```
and _are_ used by

TinaAMD commented on lib/Conversion/TosaToXTenNN.cpp:
```diff
@@ -0,0 +1,265 @@
+    if (!quantizeMulOp->hasOneUse() || !dequantizeMulOp->hasOneUse()) {
+      return rewriter.notifyMatchFailure(
+          dequantizeMulOp->getLoc(),
+          "multiplications around the QDQ operations must have single uses.");
```
"a single user"?

TinaAMD commented on lib/Conversion/TosaToXTenNN.cpp:
```diff
@@ -0,0 +1,265 @@
+
+    // Remove the multiplication around the quantize
+    rewriter.replaceOp(dequantizeMulOp, newDQ->getResults());
+
```
You do not clean up the now unused mul+quant, but did this in `CastsToQDQOps` with `quantizeOp`. I am fine with both, let DCE clean it up or do it manually (but consistent).

TinaAMD commented on test/Conversion/TosaToXTenNN/quantization.mlir:
```diff
@@ -0,0 +1,354 @@
+// CHECK:             %[[VAL_3:.*]] = "tosa.mul"(%[[VAL_2]], %[[VAL_1]]) {shift = 0 : i32} : (tensor<1x3x4x4xf32>, tensor<1x1x1x1xf32>) -> tensor<1x3x4x4xf32>
+// CHECK:             return %[[VAL_3]] : tensor<1x3x4x4xf32>
+// CHECK:           }
+  func.func @mul_missing_quantize(%arg0: tensor<1x3x4x4xsi8>) -> tensor<1x3x4x4xf32> {
```
Is this a duplicate?

cmcgirr-amd replied:
yes :D will remove

TinaAMD reviewed (commented):
Nice testing!

cmcgirr-amd commented:
As mentioned today, I will refactor this conversion to better handle constant -> Q -> DQ patterns. Taking folding casts implementations from the other feature branch on Xilinx/llvm-project.

cmcgirr-amd commented:
> As mentioned today, I will refactor this conversion to better handle constant -> Q -> DQ patterns. Taking folding casts implementations from the other feature branch on Xilinx/llvm-project.

This work though will be done in another PR, so we can get the changes we have now in.

Commit:
refactor(TosaToXTenNN.cpp): cast check to one function

Commit:
style(TosaToXTenNN.cpp): align rewrite pattern

use DCE to remvoe dead operations rather than explicitly

Commit:
refactor(TosaToXTenNN.cpp): move QDQ pattern check to the front

Commit:
fix(quantization.mlir): remove duplicate test case
